src/frontend/src/optimizer/plan_node/stream_vector_index_write.rs

```diff
@@ -0,0 +1,266 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::num::NonZeroU32;
+
+use itertools::Itertools;
+use pretty_xmlish::{Pretty, XmlNode};
+use risingwave_common::catalog::{
+    ColumnCatalog, ConflictBehavior, CreateType, Engine, OBJECT_ID_PLACEHOLDER, StreamJobStatus,
+    TableId,
+};
+use risingwave_common::hash::VnodeCount;
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_pb::catalog::PbVectorIndexInfo;
+use risingwave_pb::stream_plan::stream_node::PbNodeBody;
+
+use crate::TableCatalog;
+use crate::catalog::table_catalog::TableType;
+use crate::catalog::{DatabaseId, SchemaId};
+use crate::error::ErrorCode;
+use crate::optimizer::StreamOptimizedLogicalPlanRoot;
+use crate::optimizer::plan_node::derive::{derive_columns, derive_pk};
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::plan_node::generic::{GenericPlanRef, PhysicalPlanRef};
+use crate::optimizer::plan_node::stream::StreamPlanNodeMetadata;
+use crate::optimizer::plan_node::utils::{Distill, childless_record, plan_can_use_background_ddl};
+use crate::optimizer::plan_node::{
+    ExprRewritable, PlanBase, PlanNodeMeta, PlanTreeNodeUnary, Stream, StreamNode,
+    StreamPlanRef as PlanRef, reorganize_elements_id,
+};
+use crate::optimizer::property::{Cardinality, Distribution, Order, RequiredDist, StreamKind};
+use crate::stream_fragmenter::BuildFragmentGraphState;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct StreamVectorIndexWrite {
+    pub base: PlanBase<Stream>,
+    /// Child of Vector Index Write plan
+    input: PlanRef,
+    table: TableCatalog,
+}
+
+impl StreamVectorIndexWrite {
+    fn new(input: PlanRef, table: TableCatalog) -> crate::error::Result<Self> {
+        if input.stream_kind() != StreamKind::AppendOnly {
+            return Err(ErrorCode::NotSupported(
+                "cannot create vector index on non-append-only workflow".to_owned(),
+                "try create vector index on append only workflow".to_owned(),
+            )
+            .into());
+        }
+        let base = PlanBase::new_stream(
+            input.ctx(),
+            input.schema().clone(),
+            Some(table.stream_key.clone()),
+            input.functional_dependency().clone(),
+            input.distribution().clone(),
+            StreamKind::AppendOnly,
+            input.emit_on_window_close(),
+            input.watermark_columns().clone(),
+            input.columns_monotonicity().clone(),
+        );
+        Ok(Self { base, input, table })
+    }
+
+    pub fn create(
+        StreamOptimizedLogicalPlanRoot {
+            plan: input,
+            required_order: user_order_by,
```

Comment:
Is this `user_order_by` always empty?

Reply:
Probably. This field is not manually set with some special logic in this PR in the plan root, but follow the general optimizer process to generate the plan root. So here we just pass through the field to where we derive the table catalog.